python/tank/authentication/unified_login_flow2/authentication.py

```diff
@@ -0,0 +1,284 @@
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+import json
+import os
+import platform
+import time
+
+from tank_vendor import six
+from tank_vendor.six.moves import http_client, urllib
+
+from .. import errors
+from ...util.shotgun import connection
+
+from ... import LogManager
+
+logger = LogManager.get_logger(__name__)
+
+
+def process(
+    sg_url,
+    http_proxy=None,
+    product=None,
+    browser_open_callback=None,
+    keep_waiting_callback=lambda: True,
+):
+    sg_url = connection.sanitize_url(sg_url)
+
+    if not product and "TK_AUTH_PRODUCT" in os.environ:
+        product = os.environ["TK_AUTH_PRODUCT"]
+
+    assert product
+    assert callable(browser_open_callback)
+    assert callable(keep_waiting_callback)
+
+    url_handlers = [urllib.request.HTTPHandler]
+    if http_proxy:
+        proxy_addr = _build_proxy_addr(http_proxy)
+        sg_url = urllib.parse.urlparse(sg_url)
+
+        url_handlers.append(
+            urllib.request.ProxyHandler(
+                {
+                    sg_url.scheme: proxy_addr,
+                }
+            )
+        )
+
+    url_opener = urllib.request.build_opener(*url_handlers)
+
+    request = urllib.request.Request(
+        urllib.parse.urljoin(sg_url, "/internal_api/app_session_request"),
+        # method="POST", # see bellow
+        data=urllib.parse.urlencode(
+            {
+                "appName": product,
+                "machineId": platform.node(),
+            }
+        ).encode(),
+    )
+
+    # Hook for Python 2
+    request.get_method = lambda: "POST"
+
+    try:
+        response = http_request(url_opener, request)
+        if response.code != http_client.OK:
+            raise errors.AuthenticationError("Request denied", response.json)
+
+        session_id = response.json["sessionRequestId"]
+    except KeyError:
+        raise errors.AuthenticationError(
+            "Proto error - invalid response data - no sessionRequestId key"
+        )
+
+    if not session_id:
+        raise errors.AuthenticationError("Proto error - token is empty")
+
+    logger.debug("session ID: {session_id}".format(session_id=session_id))
+    try:
+        ret = browser_open_callback(
+            urllib.parse.urljoin(
+                sg_url,
+                "/app_session_request/{session_id}".format(
+                    session_id=session_id,
+                ),
+            ),
+        )
+        if not ret:
+            raise errors.AuthenticationError("Unable to open local browser")
+
+        logger.debug("awaiting browser login...")
+
+        sleep_time = 2
+        request_timeout = 180  # 5 minutes
+        request = urllib.request.Request(
+            urllib.parse.urljoin(
+                sg_url,
+                "/internal_api/app_session_request/{session_id}".format(
+                    session_id=session_id,
+                ),
+            ),
+            # method="PUT",
```

Comment:
Are these comments intentional? (same on the `finally` block).

Reply:
Yes, this is a reminder for when we drop Python 2 for real.
I had to rework some nice code into a worst one in order to make it work for both 2 and 3. Let's remove that in 9 months:)

Reply:
Can't wait to add Type Annotations to our source code.